flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducerBase.java

```diff
@@ -276,6 +314,41 @@ public void close() throws Exception {
 		checkErroneous();
 	}
 
+	// ------------------- Logic for handling checkpoint flushing -------------------------- //
+
+	private void acknowledgeMessage() {
+		if(!flushOnCheckpoint) {
+			// the logic is disabled
+			return;
+		}
+		pendingRecords--;
+	}
+
+	@Override
+	public Serializable snapshotState(long checkpointId, long checkpointTimestamp) {
+		if(flushOnCheckpoint) {
+			// flushing is activated: We need to wait until pendingRecords is 0
+			while(pendingRecords > 0) {
+				try {
+					Thread.sleep(10);
```

Comment:
The problem is that the `flush()` method is only implemented by the Kafka 0.9 producer, not by the 0.8 implementation.
As you can see from the classname, its the shared base class between the two version specific implementations. I think for the 0.8 producer, there is no way around the waiting approach.

I'll update the pull request to call `flush()` on the 0.9 producer.